omnisafe/algorithms/offline/base.py

```diff
@@ -0,0 +1,125 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Implementation of a basic algorithm framework for offline algorithms."""
+
+import time
+from abc import abstractmethod
+from typing import Tuple, Union
+
+import torch
+
+from omnisafe.adapter import OfflineAdapter
+from omnisafe.algorithms.base_algo import BaseAlgo
+from omnisafe.common.logger import Logger
+from omnisafe.common.offline.dataset import OfflineDataset
+from omnisafe.models.base import Actor
+from omnisafe.utils.config import Config
+
+
+class BaseOffline(BaseAlgo):
+    """Base class for offline algorithms."""
+
+    def __init__(self, env_id: str, cfgs: Config) -> None:
+        super().__init__(env_id, cfgs)
+
+        self._actor: Actor
+        self.epoch: int = 0
+
+    def _init(self) -> None:
+        self._dataset = OfflineDataset(
+            self._cfgs.train_cfgs.dataset,
+            batch_size=self._cfgs.algo_cfgs.batch_size,
+            device=self._device,
+        )
+
+    def _init_env(self) -> None:
+        self._env = OfflineAdapter(self._env_id, self._seed, self._cfgs)
+
+    def _init_log(self) -> None:
+        self._logger = Logger(
+            output_dir=self._cfgs.logger_cfgs.log_dir,
+            exp_name=self._cfgs.exp_name + f'-{self._cfgs.train_cfgs.dataset}',
+            seed=self._cfgs.seed,
+            use_tensorboard=self._cfgs.logger_cfgs.use_tensorboard,
+            use_wandb=self._cfgs.logger_cfgs.use_wandb,
+            config=self._cfgs,
+        )
+
+        self._logger.register_key('Metrics/EpRet')
+        self._logger.register_key('Metrics/EpCost')
+        self._logger.register_key('Metrics/EpLen')
+
+        self._logger.register_key('Time/Total')
+        self._logger.register_key('Time/Epoch')
+        self._logger.register_key('Time/Update')
+        self._logger.register_key('Time/Evaluate')
+
+        self._logger.register_key('Train/Epoch')
+        self._logger.register_key('TotalSteps')
+
+    def learn(self) -> Tuple[Union[int, float], ...]:
+        self._logger.log('Start training ...')
+
+        start_time = time.time()
+        epoch_time = time.time()
+
+        for step in range(self._cfgs.train_cfgs.total_steps):
+            batch = self._dataset.sample()
+            self._train(batch)
+
+            if (step + 1) % self._cfgs.algo_cfgs.steps_per_epoch == 0:
+                self.epoch = (step + 1) // self._cfgs.algo_cfgs.steps_per_epoch
+                self._logger.store(**{'Time/Update': time.time() - epoch_time})
+                evla_time = time.time()
+                self._evaluate()
+
+                self._logger.store(
+                    **{
+                        'Time/Evaluate': time.time() - evla_time,
```

Comment:
```suggestion
                        'Time/Evaluate': time.time() - eval_time,
```
And all related places